### PR TITLE
mon/OSDMonitor: fix osd_epochs insertion on prepare_beacon()

### PR DESCRIPTION
**Pull request by Matan-B**
Previously, an OSD marked out could have been added back to the `osd_epochs` map upon new beacon sent by the OSD.
If the OSD goes down (after being added back to map) - it won't be marked as OUT again.
Therefore, `osd_epochs` would remain with a "stale" OSD, as a result `OSDMonitor::get_min_last_epoch_clean()` will always return the "stale" OSD epoch.

---

When osd is marked out, it will be erased from `osd_epochs` map.
Correspondingly, it should be added back to the map **only** once it's marked in.

Fixes: https://tracker.ceph.com/issues/63883

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by ronen-fr** [comment]
Seems simple enough... fast work on that bug!

**Review by ronen-fr — APPROVED** [review]

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by athanatos** [comment]
It's not clear to me why that structure depends on an OSD being marked in or out -- an out OSD can actually be the primary for a PG via pg_temp.  e62269c8 seems to have introduced an is_out check, but the commit message actually mentions down.  I think e62269c8 meant to condition it on is_down (which would also be wrong, should have been is_up). In a larger sense, I don't actually understand the loop in get_min_last_epoch_clean -- we record per-pg last-epoch-clean values so it doesn't seem like OSD beacons should be relevant at all.  I think we need an answer to that last question in order to decide how to correct this bug.

**Comment by ljflores** [comment]
@Matan-B want this to go into qa? Just noticing @athanatos 's comment.

**Comment by NUABO** [comment]
@Matan-B @athanatos @ljflores  I would like to ask if there are still any questions about this patch? And will we incorporate it into the next version? thanks

**Comment by Matan-B** [comment]
> It's not clear to me why that structure depends on an OSD being marked in or out -- an out OSD can actually be the primary for a PG via pg_temp. [e62269c](https://github.com/ceph/ceph/commit/e62269c8929e414284ad0773c4a3c82e43735e4e) seems to have introduced an is_out check, but the commit message actually mentions down. I think [e62269c](https://github.com/ceph/ceph/commit/e62269c8929e414284ad0773c4a3c82e43735e4e) meant to condition it on is_down (which would also be wrong, should have been is_up).

Fixed the check to erase down OSDS from osd_epochs and assert that the relevant OSD is up.

> In a larger sense, I don't actually understand the loop in get_min_last_epoch_clean -- we record per-pg last-epoch-clean values so it doesn't seem like OSD beacons should be relevant at all. I think we need an answer to that last question in order to decide how to correct this bug.

I renamed `get_lower_bound` to `get_lower_bound_by_pool`. It seems that in `get_min_last_epoch_clean` we first get the lec from the **existing**/(not removed)pool reports and then scan OSD epochs. IIUC, the intent here is to avoid relying on existing pools only for trimming. For instance, if all pools were removed, `get_lower_bound_by_pool` will return the current OSDMap - which in turn will allow trimming up to it.
Skpping the OSD epochs scan is (probably) safe since lec should be determined by each existing pg. However, I'm assuming it is there as a precaution.

Left the previous commit for context.

**Comment by NUABO** [comment]
@athanatos @rzarzynski  please review :)

File: src/mon/OSDMonitor.h [hunk 1]
```diff
@@ -114,7 +114,7 @@ class LastEpochClean {
 public:
   void report(unsigned pg_num, const pg_t& pg, epoch_t last_epoch_clean);
   void remove_pool(uint64_t pool);
-  epoch_t get_lower_bound(const OSDMap& latest) const;
```
Comment (athanatos) [inline src/mon/OSDMonitor.h:117]:
```
/**
 * get_lower_bound_by_pool
 *
 * Returns epoch e such that e <= pg.last_epoch_clean for all pgs in cluster.  May
 * return 0 if any pool does not have comprehensive values for all pgs.
 */
```

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -2287,11 +2282,12 @@ version_t OSDMonitor::get_trim_to() const
 
 epoch_t OSDMonitor::get_min_last_epoch_clean() const
 {
-  auto floor = last_epoch_clean.get_lower_bound(osdmap);
+  auto floor = last_epoch_clean.get_lower_bound_by_pool(osdmap);
   // also scan osd epochs
   // don't trim past the oldest reported osd epoch
```
Comment (athanatos) [inline src/mon/OSDMonitor.cc:2287]:
```
/* There are two constraints on trimming:
 * 1. we must not trim past the last_epoch_clean for any pg
 * 2. we must not trim past the last reported epoch for any up
 *    osds.
 *
 * LastEpochClean::get_lower_bound_by_pool gives a value <= constraint 1.
 * For constraint 2, we take the min over osd_epochs, which is populated with
 * MOSDBeacon::version, see OSDMonitor::prepare_beacon
 */
```

**Review by athanatos — APPROVED** [review]
As a rule, when we fix a bug related to a misunderstanding of what a member is used for/what the invariants are, it's worth adding comments to ensure that future readers won't face the same ambiguity.  In addition to my other two comments, let's add the following to the osd_epochs declaration:

```
  /**
    * osd_epochs
    *
    * Records the MOSDBeacon::version (the osd epoch at which the OSD sent the
    * beacon) of the most recent beacon recevied from each currently up OSD.
    * Used in OSDMonitor::get_min_last_epoch_clean().
    * Down osds are trimmed upon commit of each map 
    *  (OSDMonitor::update_from_paxos).
    *  
    * TODO: use last_osd_report to store the osd report epochs, once we don't
    * need to upgrade from pre-luminous releases.
    */
  std::map<int,epoch_t> osd_epochs;
```

**Comment by amathuria** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomc58uc37eU1955-wip-yuri4-testing-2024-02-13-1546

**Comment by dvanders** [comment]
> It's not clear to me why that structure depends on an OSD being marked in or out -- an out OSD can actually be the primary for a PG via pg_temp. [e62269c](https://github.com/ceph/ceph/commit/e62269c8929e414284ad0773c4a3c82e43735e4e) seems to have introduced an is_out check, but the commit message actually mentions down. I think [e62269c](https://github.com/ceph/ceph/commit/e62269c8929e414284ad0773c4a3c82e43735e4e) meant to condition it on is_down (which would also be wrong, should have been is_up). In a larger sense, I don't actually understand the loop in get_min_last_epoch_clean -- we record per-pg last-epoch-clean values so it doesn't seem like OSD beacons should be relevant at all. I think we need an answer to that last question in order to decide how to correct this bug.

@athanatos that is_out check was a bug which we found quickly and fixed in the subsequent nautilus point release, see https://github.com/ceph/ceph/pull/36977

Then there was some cleanup to that loop in f994925908c720f8e0b3fda2f1fc51ef6e757de3

Is it clear to anyone when we regressed again, not cleaning osdmaps?

**Comment by athanatos** [comment]
@dvanders I'm not sure I understand the question.

https://github.com/ceph/ceph/pull/36977/files swapped the sense from is_in to is_out.  This is certainly less incorrect in that down osds are often marked out eventually, but it's still the wrong axis.  Because the clear condition was the in->out transition, an out osd was still able to send a beacon and get added back into map.  When it was eventually stopped, the up->down transition did not cause it to be pruned resulting in it being in the map indefinitely.

TLDR: I think https://github.com/ceph/ceph/pull/36977 only sort of fixed some of the bug.

File: src/mon/OSDMonitor.cc [hunk 3]
```diff
@@ -906,12 +906,7 @@ void OSDMonitor::update_from_paxos(bool *need_bootstrap)
       if (state & CEPH_OSD_UP) {
 	// could be marked up *or* down, but we're too lazy to check which
 	last_osd_report.erase(osd);
-      }
-    }
-    for (auto [osd, weight] : inc.new_weight) {
-      if (weight == CEPH_OSD_OUT) {
-        // manually marked out, so drop it
-        osd_epochs.erase(osd);
+	osd_epochs.erase(osd);
```
Comment (trociny) [inline src/mon/OSDMonitor.cc:909]: @Matan-B @athanatos Sorry for a late comment. We just had a customer that had stepped on this issue and noticed this PR, and now I am trying to understand the fix. And the comment above ("could be marked up *or* down, but we're too lazy to check which") makes we wonder won't we have cases when an osd is removed from `osd_epochs` when it is actually up?

Comment (Matan-B) [reply src/mon/OSDMonitor.cc:909]: OSDMonitor marks OSD as down by updating its pending_inc.new_state to CEPH_OSD_UP, grep for `= CEPH_OSD_UP`. I agree this is confusing and probably worth revisiting out of scope of this PR.

Comment (trociny) [reply src/mon/OSDMonitor.cc:909]: I see that e.g. in `OSDMap::Incremental::get_net_marked_down` when checking for `CEPH_OSD_UP` flag, it also checks the previous state to decide if it is marked down or up.

Anyway, I just made this experiment:
```
ceph osd down 0 
```
The mon set it down, the osd detected it is down in the osdmap and "reboot" and became up again in the osdmap, but now `ceph report` shows that the osd.0 is missing in `osd_epochs`. Triggering mon reelection makes the osd.0 back in `osd_epoch`.

Comment (Matan-B) [reply src/mon/OSDMonitor.cc:909]: Marking an OSD down is not the same as actually shutting it down.

When marking the OSD down, it will get erased from osd_epochs. However, the OSD is still running and continues to send heartbeats. On the next heartbeat/beacon sent, the OSD will be inserted back to osd_epochs, regardless of mon's reelection.
You can test this by setting a short `osd_beacon_report_interval`.
If I didn't understand you correctly, please submit a tracker and we can continue the discussion there, Thanks!

Comment (trociny) [reply src/mon/OSDMonitor.cc:909]: Ah, indeed the osd was reinserted eventually. I was not aware the default value for `osd_beacon_report_interval` was so large (5 minutes). I am still not sure though the behavior is entirely correct, though the issue seems very minor now. I will think if I want to open a ticket for this.

Comment (Matan-B) [reply src/mon/OSDMonitor.cc:909]: Sounds good! 
btw, issues around `osd_epochs` can be fixed by simply restarting the leader mon since "stale" data will be removed on restart.

Comment (NUABO) [reply src/mon/OSDMonitor.cc:909]: > Ah, indeed the osd was eventually reinserted. I didn't know the default was so big (5 minutes). I'm still not sure if the behavior is entirely correct, although the issue seems minor right now. I'd consider whether or not to issue a ticket for this.`osd_beacon_report_interval`

hi @sawdust, How much could you change osd_beacon_report_interval to reproduce the problem?
